src/main/java/de/zalando/aruha/nakadi/service/EventPublisher.java

```diff
@@ -0,0 +1,140 @@
+package de.zalando.aruha.nakadi.service;
+
+import de.zalando.aruha.nakadi.domain.BatchItem;
+import de.zalando.aruha.nakadi.domain.BatchItemResponse;
+import de.zalando.aruha.nakadi.domain.EventPublishResult;
+import de.zalando.aruha.nakadi.domain.EventPublishingStatus;
+import de.zalando.aruha.nakadi.domain.EventPublishingStep;
+import de.zalando.aruha.nakadi.domain.EventType;
+import de.zalando.aruha.nakadi.exceptions.EventPublishingException;
+import de.zalando.aruha.nakadi.exceptions.EventValidationException;
+import de.zalando.aruha.nakadi.exceptions.InternalNakadiException;
+import de.zalando.aruha.nakadi.exceptions.NoSuchEventTypeException;
+import de.zalando.aruha.nakadi.exceptions.PartitioningException;
+import de.zalando.aruha.nakadi.partitioning.PartitionResolver;
+import de.zalando.aruha.nakadi.repository.TopicRepository;
+import de.zalando.aruha.nakadi.repository.db.EventTypeCache;
+import de.zalando.aruha.nakadi.validation.EventTypeValidator;
+import de.zalando.aruha.nakadi.validation.ValidationError;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+public class EventPublisher {
+    private static final Logger LOG = LoggerFactory.getLogger(EventPublisher.class);
+
+    private final TopicRepository topicRepository;
+    private final EventTypeCache cache;
+    private final PartitionResolver partitionResolver;
+
+    public EventPublisher(final TopicRepository topicRepository,
+                         final EventTypeCache cache,
+                         final PartitionResolver partitionResolver) {
+        this.topicRepository = topicRepository;
+        this.cache = cache;
+        this.partitionResolver = partitionResolver;
+    }
+
+    public EventPublishResult publish(JSONArray events, String eventTypeName) throws NoSuchEventTypeException,
```

Comment:
One could ask why I mixed a result object with exceptions here. For example:

1. if there is a validation partitioning or publishing errors, it returns a result object
2. if the event type cannot be found, it throws an exception

That's because in the former scenarios we want not only to inform about an error but also to provide granular information about the status of every item. So we would be required to create new exceptions types to hold the "result" information.

It seemed to be better to make a distinction between exceptions (caused by unexpected things) and a result object (expected).

This distinction sometimes blurry so if you fill we should decide between exceptions only or result only, please, I would like to read your reasoning.